core/src/main/java/com/graphhopper/routing/AlternativeRouteCH.java

```diff
@@ -0,0 +1,240 @@
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing;
+
+import com.carrotsearch.hppc.IntIndexedContainer;
+import com.carrotsearch.hppc.predicates.IntObjectPredicate;
+import com.graphhopper.routing.ch.CHWeighting;
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.storage.SPTEntry;
+import com.graphhopper.util.EdgeIteratorState;
+
+import java.util.*;
+
+/**
+ *
+ * Minimum number-of-moving-parts implementation of alternative route search with
+ * contraction hierarchies.
+ *
+ * "Alternative Routes in Road Networks" (Abraham et al.)
+ *
+ * @author michaz
+ */
+public class AlternativeRouteCH extends DijkstraBidirectionCHNoSOD {
+
+    private double maxWeightFactor = 1.4;
+    private double maxShareFactor = 0.6;
+    public static final double T = 10000.0;
+
+    public AlternativeRouteCH(Graph graph, Weighting weighting) {
+        super(graph, weighting);
+    }
+
+    List<AlternativeInfo> calcAlternatives(final int s, final int t) {
+        // First, do a regular bidirectional route search
+        checkAlreadyRun();
+        init(s, 0, t, 0);
+        runAlgo();
+        final Path bestPath = extractPath();
+        if (!bestPath.isFound()) {
+            return Collections.emptyList();
+        }
+
+        final List<AlternativeInfo> alternatives = new ArrayList<>();
+        alternatives.add(new AlternativeInfo(bestPath, 0));
+        final double maxWeight = maxWeightFactor * bestPath.getWeight();
+
+        bestWeightMapFrom.forEach(new IntObjectPredicate<SPTEntry>() {
+            @Override
+            public boolean apply(final int v, final SPTEntry fromSPTEntry) {
+                SPTEntry toSPTEntry = bestWeightMapTo.get(v);
+                if (toSPTEntry == null)
+                    return true;
+
+                // Filter alternatives that are too long
+                if (fromSPTEntry.getWeightOfVisitedPath() + toSPTEntry.getWeightOfVisitedPath() > maxWeight)
+                    return true;
+
+                // This gives us a path s -> v -> t, but since we are using contraction hierarchies,
+                // s -> v and v -> t need not be shortest paths. In fact, they can sometimes be pretty strange.
```

Comment:
IMO it is not something special for CH that `s -> v -> t` is not a shortest path. Can we instead write something like

> Please note that if there is a shortest path `s -> v` and a shortest path `v -> t` this does not mean that `v` is on the shortest path `s  -> t`.

Reply:
No, s -> v and v -> t are not shortest paths at that point, and that's the problem.

Reply:
Can you explain why this is the case? I do not understand why e.g. `s -> v` is not shortest although it was fetched from the bestWeightMapFrom.

Reply:
Because bestWeightMapFrom is the open set, not the shortest path tree. v is not closed yet, from either direction.

Reply:
You're right since _that_ will also be the case without CH. But without CH, this will _almost_ be a shortest path. With CH, it can be arbitrarily ugly.

Reply:
Because bestWeightMapFrom is the open set, not the shortest path tree. v is not necessarily closed yet, from either direction. You're right in that's also the case without CH, but _with_ CH, those paths can get arbitrarily ugly because one single shortcut can span a really circuitous route.

Reply:
Not only that. The CH fwd/bwd searches even *settle* nodes that are not reached via a shortest path. Essentially this is because the searches are restricted to only relax relax edges leading to higher (or equal) level nodes. This is also the reason why we can prune the searches with stall on demand. See e.g. ch. 3.5.2 here: https://algo2.iti.kit.edu/documents/routeplanning/geisberger_dipl.pdf

Reply:
> v is not necessarily closed yet, from either direction
> ...
> The CH fwd/bwd searches even settle nodes that are not reached via a shortest path. 

Ah, ok - thanks! I need to relearn some basics of GH and CH it seems :)

> See e.g. ch. 3.5.2 here: https://algo2.iti.kit.edu/documents/routeplanning/geisberger_dipl.pdf.

Thx! BTW: for firefox this link works :)

https://algo2.iti.kit.edu/documents/routeplanning/geisberger_dipl.pdf#subsubsection.3.5.2

Reply:
> BTW: for firefox this link works :)

>:-(

Reply:
> `>:-(`

if this is a sad Emoji I have good news: it works for chrome too ;)

Reply:
Currently the query speed is 50 times slower than the ordinary `routingCH.mean metric`, but there is room for improvement, e.g. this branch: https://github.com/graphhopper/graphhopper/compare/try_alt_ch_faster 

For this branch and on our big map the speed up is even 6x and the `routingCH_alt.alternative_rate`is above 2. So it is basically just faster without "quality reduction" (I have no idea how to measure the real quality here, so I hope the count is a good indicator for the moment to analyze speed :) )